cluster/assemblestate/selector.go

```diff
@@ -0,0 +1,380 @@
+		addresses:     bimap.New[string, addrID](),
+		verified:      &bitset.Bitset[edgeID]{},
+		known:         make(map[peerID]*bitset.Bitset[edgeID]),
+		sources:       make(map[edgeID]int),
+		authoritative: make(map[DeviceToken]edgeID),
+	}
+}
+
+type (
+	// peerID is an opaque identifier that represents a peer. This is used to
+	// intern our strings to limit memory usage.
+	peerID int
+	// edgeID is an opaque identifier that represents an edge. This is used to
+	// intern our strings to limit memory usage.
+	edgeID int
+	// addrID is an opaque identifier that represents an address. This is used
+	// to intern our strings to limit memory usage.
+	addrID int
+)
+
+type edge struct {
+	from, to peerID
+	via      addrID
+}
+
+func (m *PrioritySelector) peerID(rdt DeviceToken) peerID {
+	if pid, ok := m.rdts.IndexOf(rdt); ok {
+		return pid
+	}
+
+	pid := m.rdts.Add(rdt)
+	m.known[pid] = &bitset.Bitset[edgeID]{}
+
+	return pid
+}
+
+func (m *PrioritySelector) addrID(a string) addrID {
+	return m.addresses.Add(a)
+}
+
+func (m *PrioritySelector) edgeID(e edge) (edgeID, bool) {
+	if eid, ok := m.edges.IndexOf(e); ok {
+		return eid, true
+	}
+
+	eid := m.edges.Add(e)
+
+	return eid, false
+}
+
+// AddAuthoritativeRoute informs the selector of an authoritative route from
+// this local node to the given peer. This route can safely be published to the
+// given peer, regardless of our knowledge of that peer's identity.
+func (m *PrioritySelector) AddAuthoritativeRoute(to DeviceToken, via string) {
+	eid, _ := m.edgeID(edge{
+		from: m.peerID(m.self),
+		to:   m.peerID(to),
+		via:  m.addrID(via),
+	})
+	m.authoritative[to] = eid
+}
+
+// RecordRoutes records all give routes and marks them as known to the given [DeviceToken].
+// The provided identified function is used to verify routes and mark them as
+// safe to publish if all we know all devices involved in a route.
+func (m *PrioritySelector) RecordRoutes(source DeviceToken, r Routes, identified func(DeviceToken) bool) (added int, total int, err error) {
+	pid := m.peerID(source)
+
+	if len(r.Routes)%3 != 0 {
+		return 0, 0, errors.New("length of routes list must be a multiple of three")
+	}
+
+	for i := 0; i+2 < len(r.Routes); i += 3 {
+		if r.Routes[i] < 0 || r.Routes[i+1] < 0 || r.Routes[i+2] < 0 {
+			return 0, 0, errors.New("invalid index in routes")
+		}
+
+		if r.Routes[i] >= len(r.Devices) || r.Routes[i+1] >= len(r.Devices) || r.Routes[i+2] >= len(r.Addresses) {
+			return 0, 0, errors.New("invalid index in routes")
+		}
+
+		fromRDT := r.Devices[r.Routes[i]]
+		toRDT := r.Devices[r.Routes[i+1]]
+
+		fromID := m.peerID(fromRDT)
+		toID := m.peerID(toRDT)
+		viaID := m.addrID(r.Addresses[r.Routes[i+2]])
+
+		eid, existed := m.edgeID(edge{
+			from: fromID,
+			to:   toID,
+			via:  viaID,
+		})
+		if !existed {
+			added++
+		}
+
+		if !m.known[pid].Has(eid) {
+			m.sources[eid]++
+		}
+
+		m.known[pid].Set(eid)
+
+		if identified(fromRDT) && identified(toRDT) {
+			m.verified.Set(eid)
+		}
```

Comment:
short comment rephrasing the code vs the properties of the fields could be useful here